### PR TITLE
WalletConnect

### PR DESCRIPTION
**Pull request by AlexBHarley**
### Description

Adds the ability for Valora to connect to DApps via WalletConnect. This is the V0 of our implementation and just adds the ability to connect to (and disconnect from) DApps, as well as approve transactions. 

Additional functionality (exposing more methods, transaction decoding, off-chain data access) are coming, however we think it makes sense to get this in sooner rather than all at once later.

More information is available in the [WalletConnect](https://www.notion.so/clabsco/WalletConnect-678a3ccb0b874fea89f263d05f07ea94) document we're maintaining.

### Tested

This works in the emulator as well as a physical device. That said, there are two different ways to test:

##### CLI DApp (for emulator)

Checkout the celo-monorepo master branch and run

```
cd packages/sdk/wallets/wallet-walletconnect
yarn ts-node scripts/run-in-memory-client.ts
```

This will spit out a long string of text, this is the QR code you would normally scan, however with the emulator you can't scan things. In Valora you can navigate to Settings > Connected Apps > Scan, then click the text at the bottom which will display a box for you to enter your QR code.

##### Web DApp (for physical device)

Go to https://use-contractkit-git-refactor-connectors-alexbharley.vercel.app and click connect, you can then perform the actions you want (only send transaction is supported right now).

### Related issues

- [[WalletConnect] implement accept/revoke transaction screen](https://app.zenhub.com/workspaces/devx-sprint-board-5fd7e7014356ff000e7a5b19/issues/celo-org/celo-monorepo/6908)
- [[WalletConnect] implement accept/revoke session screen](https://app.zenhub.com/workspaces/devx-sprint-board-5fd7e7014356ff000e7a5b19/issues/celo-org/celo-monorepo/6907)
- [[WalletConnect] implement session management screen](https://app.zenhub.com/workspaces/devx-sprint-board-5fd7e7014356ff000e7a5b19/issues/celo-org/celo-monorepo/6906)
- [[WalletConnect] setup QR scanner in Valora to handle WalletConnect QRs](https://app.zenhub.com/workspaces/devx-sprint-board-5fd7e7014356ff000e7a5b19/issues/celo-org/celo-monorepo/6905)

### Backwards compatibility

N/A

### CODE REVIEW

File: packages/mobile/package.json
Comment (jmrossy) [inline packages/mobile/package.json:82]: Looks like lots of new dependencies here. Are all these necessary? I could be mistaken (or things could have changed) but I believe that polyfills for common node libs like buffer were already included in the wallet: https://github.com/celo-org/wallet/blob/main/packages/mobile/metro.config.js#L2

File: packages/mobile/src/navigator/QRNavigator.tsx
Comment (jmrossy) [inline packages/mobile/src/navigator/QRNavigator.tsx:103]: Nit: maybe instead of a new param for handleBarcodeDetected, just put this logic in the the saga?

File: packages/mobile/src/qrcode/utils.ts
Comment (jmrossy) [inline packages/mobile/src/qrcode/utils.ts:106]: Consider making this a `yield call(...` instead, it's more saga-esque 
Should work the same, just has some very subtle benefits.

File: packages/mobile/src/walletConnect/saga.ts
Comment (jmrossy) [inline packages/mobile/src/walletConnect/saga.ts:59]: This is nice because it's robust but wondering if it wouldn't be simpler and faster to just have a chainid in each network's env file?
https://github.com/celo-org/wallet/blob/main/packages/mobile/.env.mainnet

File: packages/mobile/src/walletConnect/saga.ts
Comment (jmrossy) [inline packages/mobile/src/walletConnect/saga.ts:164]: Can you clarify why the starting of the watcher was split off like this and needs a `takeLeading` instead of just starting it directly from the `walletConnectSaga` function below?

File: packages/mobile/src/walletConnect/saga.ts
Comment (jmrossy) [inline packages/mobile/src/walletConnect/saga.ts:170]: Just wondering, does this ever need to stop / be cancelled / timeout?

File: packages/mobile/src/walletConnect/saga.ts
Comment (jmrossy) [inline packages/mobile/src/walletConnect/saga.ts:248]: Nit do these need to be generator functions?

File: packages/mobile/src/walletConnect/screens/Sessions.tsx
Comment (jmrossy) [inline packages/mobile/src/walletConnect/screens/Sessions.tsx:64]: Simple handlers like this don't need to be wrapped in useCallback and there's some evidence that it actually hurts overall perf. Recommendations on this are conflicting but I'd personally recommend against it.

**Review by jmrossy — COMMENTED** [review]
Overall looks great! Just a few comments/questions when have a sec :)

File: packages/mobile/package.json
Comment (AlexBHarley) [reply packages/mobile/package.json:82]: You're right they're not. They're from when I was messing around rn-nodeify

File: packages/mobile/src/walletConnect/screens/Sessions.tsx
Comment (AlexBHarley) [reply packages/mobile/src/walletConnect/screens/Sessions.tsx:64]: Interesting, my default is to always useCallback things and only if I'm lazy skip it.

But it's a good point that really unless it's referenced further down in the tree there's no point.

File: packages/mobile/src/walletConnect/saga.ts
Comment (AlexBHarley) [reply packages/mobile/src/walletConnect/saga.ts:59]: Yeah definitely. Didn't think to do it like that

File: packages/mobile/src/walletConnect/saga.ts
Comment (AlexBHarley) [reply packages/mobile/src/walletConnect/saga.ts:161]: I want to say no... because it's just forwarding messages.

However I'm a little shaky on how the WC client is cleaned up and restored. More testing is needed on my end

**Review by isabellewei — COMMENTED** [review]
Overall looks really great! I tried to test it on my phone with the script in monorepo. It created the pairing but I couldn't get it to start a session.

File: packages/mobile/src/walletConnect/saga.ts
Comment (isabellewei) [reply packages/mobile/src/walletConnect/saga.ts]: +1

File: packages/mobile/src/navigator/NavigationService.ts
Comment (AlexBHarley) [reply packages/mobile/src/navigator/NavigationService.ts:33]: No idea, added back

File: packages/mobile/src/walletConnect/saga.ts
Comment (jeanregisser) [inline packages/mobile/src/walletConnect/saga.ts:58]:
```suggestion
        name: APP_NAME,
```

File: packages/mobile/src/walletConnect/saga.ts
Comment (jeanregisser) [inline packages/mobile/src/walletConnect/saga.ts:61]:
```suggestion
        url: WEB_LINK,
        icons: [appendPath(WEB_LINK, 'favicon.ico')],
```

File: packages/mobile/src/walletConnect/saga.ts
Comment (jeanregisser) [inline packages/mobile/src/walletConnect/saga.ts:185]: Can we remove `as any` here?

File: packages/mobile/src/walletConnect/saga.ts
Comment (jeanregisser) [inline packages/mobile/src/walletConnect/saga.ts:173]: Is the channel properly closed/gc'ed when a new session is started?

**Review by jeanregisser — COMMENTED** [review]
This is awesome, great work!! 🚀 

Before we merge this, I think we should cover this with an e2e test to ensure it continues working as expected in the future.
@MuckT could help with this I think.

File: packages/mobile/src/app/saga.ts
Comment (AlexBHarley) [reply packages/mobile/src/app/saga.ts:157]: Is `startsWith('/wc:')` good enough?

File: packages/mobile/src/walletConnect/screens/ActionRequest.tsx
Comment (AlexBHarley) [reply packages/mobile/src/walletConnect/screens/ActionRequest.tsx:78]: Not 100% sure what you mean, you don't want the new line on line 77?

If so, I've removed it.

File: packages/mobile/src/walletConnect/reducer.ts
Comment (AlexBHarley) [reply packages/mobile/src/walletConnect/reducer.ts:9]: That's a good question. We do want to store our own state and not rely on everything being kept in the WalletConnect client. That said I can retrieve some of them on startup, how can I mark something as "non persisted".

If you receive a request, kill the app, and start it again. You should probably be sent back to the confirmation screen no?

That's not how it works currently but I can make that change.

File: packages/mobile/src/walletConnect/saga.ts
Comment (AlexBHarley) [reply packages/mobile/src/walletConnect/saga.ts:185]: I had issues with the types here, this seemed the least invasive way to go about it. 

We still know further on that `client` is a `WalletConnectClient`. Not sure that `call` plays so well with static functions. Could I also just do `yield WalletConnectClient.init(...)`?

File: packages/mobile/src/walletConnect/saga.ts
Comment (AlexBHarley) [reply packages/mobile/src/walletConnect/saga.ts:176]: I think so? I believe the API is to return a cleanup function which I do. 

If that's not enough though I'm happy to look into it.

File: packages/mobile/src/walletConnect/reducer.ts
Comment (AlexBHarley) [reply packages/mobile/src/walletConnect/reducer.ts:41]: I personally don't think so, but if you feel strongly then I'm happy to change. We would never expect that many, and maybe it's nice to keep an array like `sessions` and `pendingSessions`.

File: packages/mobile/src/app/saga.ts
Comment (jeanregisser) [reply packages/mobile/src/app/saga.ts:157]: yes!

File: packages/mobile/src/geth/networkConfig.ts
Comment (jeanregisser) [inline packages/mobile/src/geth/networkConfig.ts:78]: I realize we have a way to read this from the geth config, see https://github.com/celo-org/wallet/blob/6148646a29e5e719df9a1b32d1580ea7e6e61be8/packages/mobile/src/geth/geth.ts#L86
Would be nice to use/expose this instead (being careful to do it only after geth init has been done).

File: packages/mobile/locales/en-US/global.json
Comment (jeanregisser) [inline packages/mobile/locales/en-US/global.json:194]:
```suggestion
  "appDescription": "A mobile payments wallet that works worldwide"
```

File: packages/mobile/src/walletConnect/screens/ActionRequest.tsx
Comment (jeanregisser) [reply packages/mobile/src/walletConnect/screens/ActionRequest.tsx:77]: I meant the space in the text after `>`: `> {t`

File: packages/mobile/src/walletConnect/saga.ts
Comment (jeanregisser) [inline packages/mobile/src/walletConnect/saga.ts:62]:
```suggestion
        description: i18n.t('global:appDescription'),
```

File: packages/mobile/src/walletConnect/saga.ts
Comment (jeanregisser) [reply packages/mobile/src/walletConnect/saga.ts:176]: That should be enough, but was curious if you had seen it being properly closed.

File: packages/mobile/src/walletConnect/saga.ts
Comment (jeanregisser) [reply packages/mobile/src/walletConnect/saga.ts:188]: `yield call(...)` is the way to go here, otherwise it bypasses the normal redux saga flow.
Fine to keep `as any` for now.

**Review by jeanregisser — APPROVED** [review]
Thanks, this is great! 🎉 
Left some very small comments but already approved!

File: packages/mobile/src/walletConnect/reducer.ts
Comment (jeanregisser) [reply packages/mobile/src/walletConnect/reducer.ts:9]: > That's a good question. We do want to store our own state and not rely on everything being kept in the WalletConnect client. That said I can retrieve some of them on startup, how can I mark something in the Redux store as "don't persist"?

We do this using using the `REHYDRATE` action at the reducer level.
Example: https://github.com/celo-org/wallet/blob/3f41b83b2ff9d1732bf9c1ede63188987b1b2fd6/packages/mobile/src/account/reducer.ts#L92-L101

> If you receive a request, kill the app, and start it again. You should probably be sent back to the confirmation screen no? That's not how it works currently but I can make that change.

I'm fine resetting any in progress WalletConnect session when relauching the app.

File: packages/mobile/src/geth/networkConfig.ts
Comment (AlexBHarley) [reply packages/mobile/src/geth/networkConfig.ts:78]: Yeah, agreed. Will keep it in mind but not change for now. Those calls are async so we'd have to refactor to make getting the config a function.

File: packages/mobile/src/firebase/firebase.ts
Comment (jeanregisser) [inline packages/mobile/src/firebase/firebase.ts:229]: To match the other statements:
```suggestion
        walletConnectEnabled: flags?.walletConnectEnabled ?? false,
```

File: packages/mobile/src/geth/networkConfig.ts
Comment (isabellewei) [reply packages/mobile/src/geth/networkConfig.ts:100]: should we create an issue for this :)

File: packages/mobile/src/app/selectors.ts
Comment (isabellewei) [inline packages/mobile/src/app/selectors.ts:56]: I don't think the question mark is needed, the state should always exist

File: packages/mobile/src/qrcode/QRScanner.tsx
Comment (isabellewei) [inline packages/mobile/src/qrcode/QRScanner.tsx:50]: Opening the QR scanner on a physical device results in a crash for me.
Screenshot_20210504-120542_Alfajores (dev)

**Comment by Celoqa** [comment]
Hi @AlexBHarley @isabellewei I have verified this issue using latest Android internal build V1.14.0 (1004294350) & Test Flight build V1.14.0(58) & observed following:

**Observations:**

- **For Emulator:**
- [x] User able to see "Connected Apps" section in Settings. Also user redirect to "Connected Apps" page with Scan button.
- [ ] Also we have not observed text at the bottom which will display a box for  to enter your QR code. [Attachment](https://drive.google.com/file/d/1WuhHcvVCF7xTOCPTfZM2RyM8K4gcfj-t/view?usp=sharing)

- **For Physical Device:**
- [x] When user browse https://use-contractkit-git-refactor-connectors-alexbharley.vercel.app this site & able to do transaction by scanning QR code after redirecting to app. [Attachment](https://drive.google.com/file/d/10aqp1S3Hfx6Otermbm6za4DOZm8xqF-n/view?usp=sharing)

**Verified On Devices:** iPhone SE (14.2), Redmi Note 8 (9.0)
Can you please let us know if we need to test any other scenario here? 
Thanks..!!

**Comment by AlexBHarley** [comment]
I can't see any of those attachments, however the link to test on should be https://use-contractkit.vercel.app.

**Comment by Celoqa** [comment]
Hey @AlexBHarley Can you please recheck we have given access to you for our attachments.
Thanks..!

**Comment by AlexBHarley** [comment]
@Celoqa I'm not sure I totally understand some of the feedback. There are zero connected applications in the video, and when you navigate to the connected applications page it doesn't display any. That behaviour is correct in my eyes.

**Comment by Celoqa** [comment]
Hey @AlexBHarley Can you please let us know exactly what process we need to follow to connect apps to Valora? And do you  have any list of apps which we can connect if yes then please provide it to us. So that it will be really helpful for us to test all the apps in right way:)
Thanks..!!

**Comment by AlexBHarley** [comment]
You can try use-contractkit.vercel.app and plock.fi

**Comment by Celoqa** [comment]
Hey @AlexBHarley I have verified this issue using  latest Android  internal  build & Test Flight build V1.14.2 & observed that we are not able to connect apps using "plock.fi". Also we have tested this issue for Alfajores build V1.14.2(60) (for iOS) & V1.14.1 (for Android) & not able to connect apps.
**Verified On Devices:** iPhone SE *(13.6.1), iPhone 6+ (12.4.5), Redmi Note 8 (9.0)
**Attachment:** [NotAbleToConnectAppsUsing"plock.fi".MP4](https://drive.google.com/file/d/1LBH6iwoUQ-k0OXeIDpPMqyQ1f5s9TOW_/view?usp=sharing)
Can you please let us know any other way to test this ?
Thanks..!!

**Comment by AlexBHarley** [comment]
OK so that's because Alfajores Valora connects to `walletconnect.celo-networks-dev.org` and Valora connects to `walletconnect.celo.org`. Both Plock and use-contractkit use the same library for connecting, so I updated them both and they should work.

You have to ensure the network you're using in the website is correct for the version of Valora you're using.

However, Plock not working with Valora shouldn't block the release. Plock is its own independent project and things very likely might be broken there. use-contractkit.vercel.app should be working.

**Comment by Celoqa** [comment]
Hey @AlexBHarley It's seems like some confusion in understanding. We request you that to avoid confusion could you please provide video/attachment for exact process to connect apps & verify it from Setting page that apps are connected. It would be really helpful for us to test it thoroughly from our end:)
Thanks..!!

**Comment by Celoqa** [comment]
Hey @AlexBHarley I have verified this issue using  latest released Android  internal  build & Test Flight build V1.14.3 & observed following:
**Observations:** 

- [x] User is able to connect apps to valora using plock.fi by scanning QR code using "use-contractkit.vercel.app" 
**Attachment:** [ConnectPlocksToValora.mp4](https://drive.google.com/file/d/17YJCnu_2j27FHYhapz9T-LHLrFZ072B4/view?usp=sharing)
**Verified On Devices:** Redmi Note 8(9.0), iPhone 6+ (12.4.5)
Thanks...!!

**Comment by AlexBHarley** [comment]
@Celoqa that video looks perfect to me. After connecting as you have done so, you can try sending transactions from plock.fi or use-contractkit.vercel.app and confirm they succeed.

Another point would be to try connecting to these applications from your mobile device. So go to use-contractkit.vercel.app on your mobile browser and try connect to Valora.

**Comment by Celoqa** [comment]
Sure @AlexBHarley We will try as per you above request & will get back to you with an update.

**Comment by Celoqa** [comment]
Hey @AlexBHarley I have again verified this issue using latest Android internal build V1.15.0 and Test Flight 1.15.0 & observed following:
**Observations:**

- [x] **For Web:** User is able to connect apps using web & able to do transaction after successfully connection. [Attachment](https://drive.google.com/file/d/1KmYcru5UaiuL40dWIjeuIk5srOU1Wl7o/view?usp=sharing)
- [ ] **For Android:** User is able to see Blank screen when trying to connect by selecting Valora application even app is installed.[Attachment](https://drive.google.com/file/d/148NijEzxrO2aG2w4AmYwVSRnU4SsjAb0/view?usp=sharing)
- [ ] **For iOS:** User get "Safari cannot open the page because the address is invalid" popup when trying to connect by selecting Valora application even app is installed.[Attachment](https://drive.google.com/file/d/12f4pMo-Qh81ILhl9UjV3mVbrAHUH7Zke/view?usp=sharing)
**Verified On Devices/Browsers:** Redmi Note  8 (9.0) , iPhone 6+ (12.4.5), Chrome Version 91.0.4472.77
Can you please let us know needful for above unchecked points? Are they expected or we need to raise tickets for same ?
Thanks..!!

**Pull request merged**
